### ISSUE TITLE
Getting "Workspaces can only be enabled in private projects" even for a private project

### ISSUE THREAD

**Issue by kaiyoma** [yarnpkg/yarn#8580]
I'm using yarn + lerna in a mono-repo, and today, after rebuilding my lockfile, I started seeing this warning from yarn:

```
[3/4] Linking dependencies...
warning Workspaces can only be enabled in private projects.
```

I'm not sure why I'm seeing this because I have this in my package.json:

```
  "private": true,
```

Seems like a bug in yarn?

Node 14.15.0
yarn 1.22.5
lerna 3.22.1

**Comment by gabegorelick** [comment]
I'm seeing this warning in a public project that does not use workspaces.

**Comment by rally25rs** [comment]
Could you provide a minimal package.json to reproduce the issue?

**Comment by IanMayo** [comment]
Possible duplicate of: https://github.com/webpack/webpack-cli/issues/479

**Comment by bkglobal** [comment]
In package.json add the directory in workspace like
```
"workspaces": [
      "app/*"
]
```

**Comment by gabegorelick** [comment]
> Possible duplicate of: webpack/webpack-cli#479

In my case, the issue is indeed caused by a dep that has `workspaces` set (although not webpack-cli in my case). 

Seems like yarn should either tell you the offending package, or just not warn at all if a dep erroneously has `workspaces` set.

**Comment by kennie-larkson** [comment]
I am having this same warning in a project that does not use workspaces upon `yarn add moment` and I have `‘private’: true` in package.json

**Comment by eps1lon** [comment]
At this point it shouldn't warn about this in dependencies at all. The dependencies might be using a package manager that uses that field differently. For example, in yarn v2 using workspaces in a published package is perfectly fine:

> Worktrees used to be required to be private (ie list "private": true in their package.json). This requirement got removed with the 2.0 release in order to help standalone projects to progressively adopt workspaces (for example by listing their documentation website as a separate workspace).

-- https://yarnpkg.com/features/workspaces#how-to-declare-a-worktree

For us, we get the warning since we have `react-redux` installed. But `react-redux` is setup correctly from their viewpoint. So the warning is not actionable.

**Comment by kaiyoma** [comment]
See also https://github.com/reduxjs/react-redux/issues/1815

**Comment by benquarmby** [comment]
[NPM says hello](https://github.com/npm/cli/blob/7a858277171813b37d46a032e49db44c8624f78f/package.json#L5-L9) and a friendly "up yours" to the only-private-packages-have-workspaces rule 👋

**Comment by ale917k** [comment]
It's been a year and a half since raised; Has anything been done or discussed about this? Can the warnings at least be omitted when flagged on dependencies? Bit annoying to have all these around..

![Screenshot 2022-07-30 at 16 40 57](https://user-images.githubusercontent.com/53540737/181919366-80dd46c8-0992-445c-8ae9-dcbaca3769de.png)

**Comment by nhhockeyplayer** [comment]
its gotten worse on node v18
can no longer install even if you toggle their property on or off

```
meanstack02@admin meanstacknh % yarn install
yarn install v1.22.19
error Workspaces can only be enabled in private projects.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
```

these institutions are in over their head and the frustration is a group failure altogether mismanaged
it all started with npmjs.org which is still collecting credit card subscriptions and nothing works like a real software project should 

hopefully microsoft will fix everything

**Comment by gduliscouet-ubitransport** [comment]
@rally25rs @merceyz  Here is a minimal reproduction package.json of the issue (just installing `semantic-release` revealed this unwanted warning):

```json
{
  "name": "yarn-warning-repro",
  "version": "1.0.0",
  "main": "index.js",
  "license": "MIT",
  "private": true,
  "dependencies": {
    "semantic-release": "^19.0.3"
  }
}
```

![image](https://user-images.githubusercontent.com/101416921/184117129-afb0dddf-cd5e-4078-a98d-98a180dc6dcf.png)

yarn version 1.22.19
node version 16.14.0

A repository of the reproduction: https://github.com/gduliscouet-ubitransport/yarn-warning-repro

**Comment by rally25rs** [comment]
thanks for the repro case @gduliscouet-ubitransport 

It looks like one of the transitive deps is `npm`

```
semantic-release > @semantic-release/npm > npm
```

The `npm` package's `package.json` indicates it is a workspace
```
{
  "version": "8.17.0",
  "name": "npm",
  "description": "a package manager for JavaScript",
  "workspaces": [
    "docs",
    "smoke-tests",
    "workspaces/*"
  ],
```

which is what is leading to this warning.

Since this is just a warning I think it can be ignored; the `npm` library still does get installed into `node_modules`

**Comment by bharathmb7** [comment]
I am getting the error
`error Workspaces can only be enabled in private projects.`
when i am trying to install the dependencies for [ens-v3](https://github.com/ensdomains/ens-app-v3)
And I am not able to proceed forward.
It can be reproduced by just cloning the above repo and running `yarn` in it.

**Comment by guillaumeduliscouet** [comment]
Thank you @rally25rs for analysing this,
Can we close this issue/subject with a clear conclusion ? I see 2 possibilities:

**1. This "workspaces should be private" rule is a standard for package managers. They should all warn (or error) if the project (or one of its dependencies) is breaking the rule.**

In this case I'll open an issue on the npm repo and any other repo breaking the rule and we can close this one.

**2. The rule is not standard, only yarn is applying it.**

In this case yarn should not warn for dependencies breaking the rule since they may use another package manager.

I agree that it is only a warning and it can be ignored, but some warnings need to be addressed and this one is just creating noise. It is encouraging devs to ignore every warnings, making warnings useless

**Comment by dylang** [comment]
This warning is typically for the a third party dependency or dependency of a dependency getting installed.

It's not likely useful for developer running the install.

This is made extra confusing by not showing the name of the module the warning is for.

My opinion is that this warning should only be displayed when publishing if a published package has `workspaces` in the `package.json`. It should not be shown at install. At this point it's too late and has no effect on the install so it doesn't matter.

The only time it should be shown at install time is if this problem is with one of the packages in the workspace. And if that's the case, it should say which package that is.

**Comment by Tofandel** [comment]
The piece of code throwing the warning

https://github.com/yarnpkg/yarn/blob/3119382885ea373d3c13d6a846de743eca8c914b/src/config.js#L873

**Comment by ijemmao** [comment]
Is there any update on this issue? I'm experiencing this issue in my public repo: https://github.com/nkowaokwu/igbo_api/actions/runs/4455179532/jobs/7824826839?pr=576

**Comment by AnonymousX86** [comment]
It is possible to explicitly disable it [^1] in `.yarnrc` file. [^2]

[^1]: https://classic.yarnpkg.com/lang/en/docs/workspaces/#toc-tips-tricks
[^2]: https://classic.yarnpkg.com/lang/en/docs/yarnrc/

**Comment by nopeless** [comment]
Is it not possible to inform the developer which project is actually offending? i.e. the workspace project that does not have `"private": true`? This seems like a useful feature

**Comment by alfaproject** [comment]
I guess no hope for this?

**Comment by natesire** [comment]
> error Workspaces can only be enabled in private projects.

It's related to the built-in @npmcli dependencies using workspaces. I have been debugging the other side of the issue from the npm side. I had to switch to yarn just to work around things. NPM CLI is broken on windows for Node 16.14+ because of NPM.

Then I saw all of these warnings pop up related to workspaces when running yarn.

https://github.com/npm/cli/pull/6601

**Comment by rmclaughlin-nelnet** [comment]
This may seems like a dumb question, but why does yarn restrict workspaces to only private projects?

**Comment by ckhatton-transreport** [comment]
After 1.5yrs of this ticket being open, this comment above seems the only option 😓 
https://github.com/yarnpkg/yarn/issues/8580#issuecomment-1483520779

**Comment by rally25rs** [comment]
@ckhatton-transreport 

> After 1.5yrs of this ticket being open, this comment above seems the only option

Yarn classic / v1 has not been in active feature development for a couple years. Major version 4 is close to being released.
Classic/v1 only receives critical security fixes and bug fixes that would prevent it from working.

While it is an admittedly confusing message, Seeing a warning does not prevent anyone from continuing to use yarn.

---

@rmclaughlin-nelnet 

> What is the logic behind restricting workspaces to only private projects?

I didn't write that feature initially, but I believe the thought behind "workspaces" was that they are a collection of deploy-able npm packages (each package in the workspace would be `npm publish`ed as a library), but they themselves are not supposed to be published libraries. I suspect the thinking behind the `private: true` check was that having it a private package would help prevent the workspace root from being `npm publish`ed as a dependency.

---

I'm going to mark this issue as `closed` because 
* yarn v1 is unlikely to release a fix for this non-breaking issue
* the message is a warning that can be suppressed in `.yarnrc`

I'd also like to take this moment to encourage folks to upgrade to yarn modern where possible.

**Comment by avarun42** [comment]
> Since this is just a warning I think it can be ignored; the `npm` library still does get installed into `node_modules`

This is useless and unproductive. Of course a warning can be ignored, but if a warning is not providing any actionable information whatsoever (you would agree that a yarn user using `npm` as a transitive dependency is not making a mistake, correct?) then that warning should not exist. Especially since the warning currently **doesn't even state which package it's referring to**.

I would consider this a bug in yarn v1.

> I'm going to mark this issue as closed because [...] the message is a warning that can be suppressed in .yarnrc

Not actually true. The mentioned `yarnrc` config **turns off workspaces entirely**. This is obviously not a solution that works if you're actually using workspaces. However, in that situation the warning message being thrown is both useless and a red herring. You can be inside a private project with workspaces enabled and end up seeing the message due to a transitive dependency.

---

Keep in mind this problem doesn't just affect users of `yarn` once. It affects every single installer of a project that's using `yarn`, who may wonder why they're seeing a warning only to end up on this thread that's telling them to ignore it — by definition a waste of time and productivity for everybody that could be solved by simply not showing a useless warning.

**Comment by jonkoops** [comment]
This bug should be re-opened and fixed, it is an issue in Yarn v1, since it is still the default, even when initializing a new project. This is actively causing issues for users of `classnames` (https://github.com/JedWatson/classnames/pull/329#issuecomment-1871740033).

**Comment by malininss** [comment]
Why is there no reaction if this is obviously a bug? I use workspaces in my project and can't turn them off in `.yarnrc`

**Comment by darklight9811** [comment]
Also think this still a bug that should be fixed

**Comment by mjugger** [comment]
this is not even a warning it is throwing an error for me and preventing the project from being installed. is there a way to completely bypass this?

**Comment by fulldecent** [comment]
Here is the solution to this issue (not a workaround). The issue should stay closed because it is solved.

**tl;dr You are running and old version of Yarn. And when you think you are upgrading it you are not.**

---

# Root cause

You are running an old version of Yarn. And when you think you are upgrading Yarn, you are not.

Proof:

```sh
yarn --version
```

Compare to [actual current Yarn releases](https://github.com/yarnpkg/berry/releases).

# Things that don't work ❌ 

Don't do these things. Intuitively they look like they will work but actually they do nothing.

- ```sh
  brew upgrade yarn
  ```

  This installs an old version. No deprecation warning. No hints. Just wrong output.

  You can see the formula at https://github.com/Homebrew/homebrew-core/blob/d01418781a024b63d9ba9e1299d7f1edf703558b/Formula/y/yarn.rb which is wrong. It mentions version 2 as the implied latest. But actually the latest is currently 4. [PR sent](https://github.com/Homebrew/homebrew-core/pull/182963).

- Add your notes for other package managers here…

# Solution

Uninstall Yarn on your system if you installed it using any non-preferred (even retroactively non-preferred) ways.

```sh
brew uninstall yarn
```

Then install the preferred way. (Here is how I prefer, there are others.)

```sh
brew install nvm
nvm install --lts --reinstall-packages-from=current # If you don't use that second flag you get old version
nvm use --lts
```

Then [install yarn](https://yarnpkg.com/getting-started/install)

```sh
corepack enable # You already have this command installed when you installed Node.js
```

Then delete your .lock file which may be preventing upgrades

```sh
rm yarn.lock
```

Now you can use yarn

```sh
yarn install
yarn set version stable # This will update your package.json to add a $.packageManager entry
```

Notice how everything looks different? You are on the new version now. And the warning is gone.

---

# Follow on work

*Please feel free to ignore any of this. Good vibes only.*

Here's how I think the world can be a better place:

- Brew should not imply that the latest version of yarn is 2.x [PR sent](https://github.com/Homebrew/homebrew-core/pull/182963)
- Yarn project should publish a page https://yarnpkg.com/getting-started/upgrade that lists all the error messages you could get from running an old version of Yarn, then all the steps above on how to upgrade, then we can all link to that page, yay!
- Yarn project should be more agressive on ending version support. Currently the Yarn project actively maintains over 100 [versions of](https://github.com/yarnpkg/yarn/releases) [the product](https://github.com/yarnpkg/berry/releases). Yes, if one version of Yarn installs another version of Yarn and does warn you to upgrade… that counts as actively supported.
- Yarn project should make one more derprecation release for each major branch, because SemVer.
- Brew should publish that new 1.x.y+1 release. Now everybody using yarn installed by Brew will get the upgrade message.

**Comment by markedwards** [comment]
@fulldecent this issue is not about yarn berry, its about yarn v1.  Please see the above discussion.

**Comment by benquarmby** [comment]
:trollface: Yarn Berry, the Node ~package manager~ module system replacement nobody asked for.

Another alternative is switching to [pnpm](https://pnpm.io/), which just keeps delivering on the original promise of Yarn 1.x; a faster more stable NPM. None of these package managers are perfect. PNPM has its own issues just like yarn 1.x, but it's actively maintained, and this particular warning does not happen. `pnpm import ./yarn.lock` is a great place to start.

**Comment by Tofandel** [comment]
It's also not just brew, if v1 really should not be installed then it shouldn't be the default in the distros package managers. Right now `apt install yarn` on any distro will get you v1 by default

If they really want people to upgrade to v4 then it shouldn't be so painfully hard to do so
